src/tensor_ops/matmul/cuda_kernel.rs

```diff
@@ -1,13 +1,150 @@
-use crate::{shapes::*, tensor::Cuda};
+use crate::{
+    shapes::*,
+    tensor::cuda::{Cuda, CudaArray},
+};
+
+use cudarc::{
+    cublas::{
+        result::CublasError, sys::cublasOperation_t, CudaBlas, Gemm, GemmConfig,
+        StridedBatchedConfig,
+    },
+    driver::{DevicePtr, DevicePtrMut},
+};
+use std::sync::Arc;
+
+/// sgemm helper.
+///
+/// # case 1: c is not transposed
+/// note that lhs becomes b and rhs becomes a since cuda expects
+/// column major data, but dfdx uses row major. So this function actually
+/// calls the underlying cuda gemm call as `(N, K) * (K, M) = (N, M)`.
+///
+/// Since `(N, K)` in column major format is equal to (K, N) in row major format,
+/// we can just pass rhs normally. We can pass the row major format `out` directly
+/// in without transposing or anything because of this fact also.
+///
+/// # case 2: c is transposed
+///
+/// lhs is a and rhs is b, but we have to transpose them if they are not already
+#[allow(clippy::too_many_arguments)]
+unsafe fn sgemm<
+    M: Dim,
+    K: Dim,
+    N: Dim,
+    A: DevicePtr<f32>,
+    B: DevicePtr<f32>,
+    C: DevicePtrMut<f32>,
+>(
+    blas: &CudaBlas,
+    m: M,
+    k: K,
+    n: N,
+    lhs: &A,
+    lhs_strides: [usize; 2],
+    rhs: &B,
+    rhs_strides: [usize; 2],
+    beta: f32,
+    out: &mut C,
+    out_strides: [usize; 2],
+) -> Result<(), CublasError> {
+    if out_strides[0] > 1 {
```

Comment:
This function assumes one of the strides is 1, should add checks to ensure this.